From b3f4fa30590d7a84cb6f4c64cbcfb1c8abeab405 Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Tue, 25 Nov 2014 16:33:23 -0500
Subject: [security] add flags to rename hooks

Message-id: <20141125163323.18115.3021.stgit@warthog.procyon.org.uk>
Patchwork-id: 100907
O-Subject: [PATCH 07/47] security: add flags to rename hooks
Bugzilla: 985875
RH-Acked-by: Ian Kent <ikent@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

From: Miklos Szeredi <mszeredi@suse.cz>

Add flags to security_path_rename() and security_inode_rename() hooks.

Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
Reviewed-by: J. Bruce Fields <bfields@redhat.com>
Original-commit: 0b3974eb04c4874e85fa1d4fc70450d12f28611d
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/cachefiles/namei.c b/fs/cachefiles/namei.c
index 92103c7..905ca65 100644
--- a/fs/cachefiles/namei.c
+++ b/fs/cachefiles/namei.c
@@ -391,7 +391,7 @@ try_again:
  path.dentry = dir;
  path_to_graveyard.mnt = cache->mnt;
  path_to_graveyard.dentry = cache->graveyard;
- ret = security_path_rename(&path, rep, &path_to_graveyard, grave);
+ ret = security_path_rename(&path, rep, &path_to_graveyard, grave, 0);
  if (ret < 0) {
   cachefiles_io_error(cache, "Rename security error %d", ret);
  } else {
diff --git a/fs/namei.c b/fs/namei.c
index d693cd1..146fa81 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -4023,7 +4023,8 @@ int vfs_rename(struct inode *old_dir, struct dentry *old_dentry,
    return error;
  }
 
- error = security_inode_rename(old_dir, old_dentry, new_dir, new_dentry);
+ error = security_inode_rename(old_dir, old_dentry, new_dir, new_dentry,
+          flags);
  if (error)
   return error;
 
@@ -4178,7 +4179,7 @@ retry_deleg:
   goto exit5;
 
  error = security_path_rename(&oldnd.path, old_dentry,
-         &newnd.path, new_dentry);
+         &newnd.path, new_dentry, flags);
  if (error)
   goto exit5;
  error = vfs_rename(old_dir->d_inode, old_dentry,
diff --git a/include/linux/security.h b/include/linux/security.h
index 55fe297..ca2b652 100644
--- a/include/linux/security.h
+++ b/include/linux/security.h
@@ -1781,7 +1781,8 @@ int security_inode_mkdir(struct inode *dir, struct dentry *dentry, umode_t mode)
 int security_inode_rmdir(struct inode *dir, struct dentry *dentry);
 int security_inode_mknod(struct inode *dir, struct dentry *dentry, umode_t mode, dev_t dev);
 int security_inode_rename(struct inode *old_dir, struct dentry *old_dentry,
-     struct inode *new_dir, struct dentry *new_dentry);
+     struct inode *new_dir, struct dentry *new_dentry,
+     unsigned int flags);
 int security_inode_readlink(struct dentry *dentry);
 int security_inode_follow_link(struct dentry *dentry, struct nameidata *nd);
 int security_inode_permission(struct inode *inode, int mask);
@@ -2149,7 +2150,8 @@ static inline int security_inode_mknod(struct inode *dir,
 static inline int security_inode_rename(struct inode *old_dir,
       struct dentry *old_dentry,
       struct inode *new_dir,
-      struct dentry *new_dentry)
+      struct dentry *new_dentry,
+      unsigned int flags)
 {
  return 0;
 }
@@ -2940,7 +2942,8 @@ int security_path_symlink(struct path *dir, struct dentry *dentry,
 int security_path_link(struct dentry *old_dentry, struct path *new_dir,
          struct dentry *new_dentry);
 int security_path_rename(struct path *old_dir, struct dentry *old_dentry,
-    struct path *new_dir, struct dentry *new_dentry);
+    struct path *new_dir, struct dentry *new_dentry,
+    unsigned int flags);
 int security_path_chmod(struct path *path, umode_t mode);
 int security_path_chown(struct path *path, kuid_t uid, kgid_t gid);
 int security_path_chroot(struct path *path);
@@ -2988,7 +2991,8 @@ static inline int security_path_link(struct dentry *old_dentry,
 static inline int security_path_rename(struct path *old_dir,
            struct dentry *old_dentry,
            struct path *new_dir,
-           struct dentry *new_dentry)
+           struct dentry *new_dentry,
+           unsigned int flags)
 {
  return 0;
 }
diff --git a/security/security.c b/security/security.c
index 94b35ae..01dbc15 100644
--- a/security/security.c
+++ b/security/security.c
@@ -435,7 +435,8 @@ int security_path_link(struct dentry *old_dentry, struct path *new_dir,
 }
 
 int security_path_rename(struct path *old_dir, struct dentry *old_dentry,
-    struct path *new_dir, struct dentry *new_dentry)
+    struct path *new_dir, struct dentry *new_dentry,
+    unsigned int flags)
 {
  if (unlikely(IS_PRIVATE(old_dentry->d_inode) ||
        (new_dentry->d_inode && IS_PRIVATE(new_dentry->d_inode))))
@@ -526,7 +527,8 @@ int security_inode_mknod(struct inode *dir, struct dentry *dentry, umode_t mode,
 }
 
 int security_inode_rename(struct inode *old_dir, struct dentry *old_dentry,
-      struct inode *new_dir, struct dentry *new_dentry)
+      struct inode *new_dir, struct dentry *new_dentry,
+      unsigned int flags)
 {
         if (unlikely(IS_PRIVATE(old_dentry->d_inode) ||
             (new_dentry->d_inode && IS_PRIVATE(new_dentry->d_inode))))
-- 
1.7.1